src/Table/ExpandableRow.tsx

```diff
@@ -1,33 +1,44 @@
-import { ReactElement, PureComponent } from 'react'
+import * as React from 'react';
+import { Subtract } from 'utility-types';
+import { DetailTable } from './DetailTable';
 
-interface Props {
-    cols: string[];
-    expandElement: ReactElement<HTMLDivElement>;
+export interface InjectedExpandableProps {
+    onClick(): void;
+    isExpand: boolean;
 }
 
 interface States {
     isExpand: boolean;
 }
 
-// TODO: In progress...
+const makeExpandable = <P extends InjectedExpandableProps>(Component: React.ComponentType<P>) =>
+    class MakeExpandable extends React.PureComponent<Subtract<P, InjectedExpandableProps>, States>{
+        state: States = {
+            isExpand: false,
+        };
+
+        onClick = () => {
+            this.setState(prevState => ({ isExpand: !prevState.isExpand }));
+        };
 
-export class ExpandableRow extends PureComponent<Props, States> {
+        render() {
+            return (
+                <Component {...this.props as P} onClick={this.onClick} isExpand={this.state.isExpand} />
```

Comment:
Why need `as` here?

Reply:
without `as` it's error `Type 'Readonly<{ children?: ReactNode; }> & Readonly<Pick<P, SetDifference<keyof P, "onClick" | "isExpand">>> & { onClick: () => void; isExpand: boolean; }' is not assignable to type 'P'.ts(2322)`

Reply:
Then it's because wrong typedef at props. Can fix that?

Reply:
Your way to write hoc look really complex. I think better write like this (P can extends React.Props<any> or object)
```ts
export const makeExpandable = <P extends React.Props<any>>(
  Component: React.ComponentType<P & InjectedExpandableProps>,
) =>
  class extends React.PureComponent<P, States> {
    public state: States = {
      isExpand: false,
    };

    public onClick = () => {
      this.setState(prevState => ({ isExpand: !prevState.isExpand }));
    };

    public render() {
      return (
        <Component
          {...this.props}
          onClick={this.onClick}
          isExpand={this.state.isExpand}
        />
      );
    }
  };

```

Reply:
can use function as well if don't want to have `any` or `object` 
```ts
export function makeExpandable<P>(
  Component: React.ComponentType<P & InjectedExpandableProps>,
) {
  return class extends React.PureComponent<P, States> {
    public state: States = {
      isExpand: false,
    };

    public onClick = () => {
      this.setState(prevState => ({ isExpand: !prevState.isExpand }));
    };

    public render() {
      return (
        <Component
          {...this.props}
          onClick={this.onClick}
          isExpand={this.state.isExpand}
        />
      );
    }
  };
}

```